app/uk/gov/hmrc/transitmovementsrouter/connectors/PushNotificationsConnector.scala

```diff
@@ -61,38 +73,71 @@ class PushNotificationsConnectorImpl @Inject() (httpClientV2: HttpClientV2, appC
     hc: HeaderCarrier,
     ec: ExecutionContext
   ): EitherT[Future, PushNotificationError, Unit] =
     EitherT {
       if (!appConfig.pushNotificationsEnabled) {
         Future.successful(Right(()))
       } else {
-        val url = baseUrl.withPath(pushNotificationMessageUpdate(movementId, messageId))
-
-        val requestBuilder = httpClientV2.post(url"$url")
-
-        val requestBuilderWithSource = source match {
-          case Some(src) => requestBuilder.setHeader(HeaderNames.CONTENT_TYPE -> MimeTypes.XML).withBody(src)
-          case None      => requestBuilder
-        }
-
-        requestBuilderWithSource
-          .execute[Either[UpstreamErrorResponse, HttpResponse]]
-          .map {
-            case Right(_) =>
-              Right(())
-            case Left(error) =>
-              error.statusCode match {
-                case NOT_FOUND             => Left(MovementNotFound(movementId))
-                case INTERNAL_SERVER_ERROR => Left(Unexpected(Some(error.getCause)))
-                case _                     => Left(Unexpected(Some(error.getCause)))
-              }
-          }
-          .recover {
-            case NonFatal(ex) => Left(Unexpected(Some(ex)))
-          }
+        val request = createRequest(movementId, messageId)
+          .setHeader(HeaderNames.CONTENT_TYPE -> MimeTypes.XML)
+          .withBody(sourceXml)
+        execute(request, movementId)
+
+      }
+    }
+
+  override def postJSON(movementId: MovementId, messageId: MessageId, sourceJson: JsValue)(implicit
+    hc: HeaderCarrier,
+    ec: ExecutionContext
+  ): EitherT[Future, PushNotificationError, Unit] =
+    EitherT {
+      if (!appConfig.pushNotificationsEnabled) {
+        Future.successful(Right(()))
+      } else {
+        val request = createRequest(movementId, messageId)
+          .setHeader(HeaderNames.CONTENT_TYPE -> MimeTypes.JSON)
+          .withBody(sourceJson)
+        execute(request, movementId)
+
       }
     }
+
+  override def post(movementId: MovementId, messageId: MessageId)(implicit
+    hc: HeaderCarrier,
+    ec: ExecutionContext
+  ): EitherT[Future, PushNotificationError, Unit] =
+    EitherT {
+      if (!appConfig.pushNotificationsEnabled) {
+        Future.successful(Right(()))
+      } else {
+        val request = createRequest(movementId, messageId)
+        execute(request, movementId)
+
+      }
+    }
+
+  private def execute(requestBuilder: RequestBuilder, movementId: MovementId)(implicit ec: ExecutionContext) =
+    requestBuilder
+      .execute[Either[UpstreamErrorResponse, HttpResponse]]
+      .map {
+        case Right(_) =>
+          Right(())
+        case Left(error) =>
+          error.statusCode match {
+            case NOT_FOUND             => Left(MovementNotFound(movementId))
+            case INTERNAL_SERVER_ERROR => Left(Unexpected(Some(error.getCause)))
+            case _                     => Left(Unexpected(Some(error.getCause)))
```

Comment:
Few things

* `UpstreamErrorResponse` is already a throwable. Just use that.
* Related, I suspect we're doing this elsewhere and you've copied from this, but it's quite possible there will be no cause in the exception, and so you'll end up with `Some(null)`, which we do not want.
* Why have two branches do the same thing, especially as one is the default? I have seen this elsewhere, but that doesn't mean it's right to do.

Reply:
I just used whatever already there with this class not copied from anywhere. I will change it as suggested above

Reply:
Yup, fair, didn't show up like that on the side by side diff. May as well change it while we're at it though.